File: LRU-Cache/LRUC.go
package main

type LRUCache struct {
	cap int

	record []int
	backet map[int]int //key -> value
}

func Constructor(capacity int) LRUCache {
	result := LRUCache{
		cap:    capacity,
		record: []int{},
		backet: map[int]int{},
	}
	return result
}

func (this *LRUCache) Get(key int) int {
	if v, ok := this.backet[key]; ok {
		for i := 0; i < len(this.record); i++ {
			if this.record[i] == key {
				first_p := this.record[:i]
				second_p := this.record[i+1:]
				this.record = append(this.record[:0], first_p...)
				this.record = append(this.record, second_p...)
				this.record = append(this.record, key)
			}
		}
		return v
	}
	return -1
}

func (this *LRUCache) Put(key int, value int) {
	if v := this.Get(key); v != -1 {
		this.backet[key] = value
		return
	}

	if len(this.record) == this.cap {
		delete(this.backet, this.record[0])
		this.record = append(this.record[:0], this.record[1:]...)
	}
	this.record = append(this.record, key)
	this.backet[key] = value
}

/**
 * Your LRUCache object will be instantiated and called as such:
 * obj := Constructor(capacity);
 * param_1 := obj.Get(key);
 * obj.Put(key,value);
 */
func main() {
	obj := Constructor(2)
	// obj.Put(1, 1)
	// obj.Put(2, 2)
	// fmt.Printf("%+v\n", obj)

	// fmt.Printf("%+v\n", obj.Get(1))
	// obj.Put(3, 3)
	// fmt.Printf("%+v\n", obj)

	obj.Get(2)
	obj.Put(2, 6)
	obj.Get(1)
	obj.Put(1, 5)
	obj.Put(1, 2)
	obj.Get(1)
	println(obj.Get(2))
}

File: LRU-Cache/LRUC.rs
use std::collections::HashMap;

#[derive(Debug)]
struct LRUCache {
    cap: i32,

    record: Vec<i32>,
    backet: HashMap<i32, i32>,
}

/**
 * `&self` means the method takes an immutable reference.
 * If you need a mutable reference, change it to `&mut self` instead.
 */
impl LRUCache {
    fn new(capacity: i32) -> Self {
        LRUCache {
            cap: capacity,
            record: vec![],
            backet: HashMap::new(),
        }
    }

    fn get(&mut self, key: i32) -> i32 {
        if let Some(v) = self.backet.get(&key) {
            let mut head = vec![key];
            let mut tail = vec![];
            //let mut find = false;
            for vv in self.record.clone() {
                if vv != key {
                    tail.push(vv);
                }
            }

            head.append(&mut tail);
            self.record = head;
            return v.clone();
        }
        -1
    }

    fn put(&mut self, key: i32, value: i32) {
        //let v = self.get(key);
        if self.get(key) != -1 {
            self.backet.insert(key, value);
            return;
        }

        let mut new = vec![key];

        if self.record.len() == self.cap as usize {
            self.backet.remove(self.record.last().unwrap());
            for i in self.record.drain(..self.record.len() - 1) {
                new.push(i);
            }
            self.record = new;
        } else {
            for i in self.record.clone() {
                new.push(i);
            }
            self.record = new;
        }
        self.backet.insert(key, value);
    }
}

/**
 * Your LRUCache object will be instantiated and called as such:
 * let obj = LRUCache::new(capacity);
 * let ret_1: i32 = obj.get(key);
 * obj.put(key, value);
 */

fn main() {
    let mut obg = LRUCache::new(2);
    println!("{}", obg.get(2));
    obg.put(1, 1);
    obg.put(2, 2);
    println!("{}", obg.get(1));
    obg.put(3, 3);
    //println!("{:#?}", obg);
    println!("{}", obg.get(2));
    obg.put(4, 4);
    println!("{}", obg.get(1));
    println!("{}", obg.get(3));
    println!("{}", obg.get(4));

    let mut obg1 = LRUCache::new(1);
    obg1.put(2, 1);
    println!("{}", obg1.get(2));
    println!("{:#?}", obg1);
    obg1.put(3, 2);
    println!("{:#?}", obg1);
    println!("{}", obg1.get(2));
    println!("{}", obg1.get(3));
}
